pkg/protocol/protocol.go

```diff
@@ -2,316 +2,216 @@ package protocol
-}
-
-func (p *Protocol) MainEngineInstance() *engine.Engine {
-	p.activeEngineMutex.RLock()
-	defer p.activeEngineMutex.RUnlock()
+// APIForVersion returns the API for the given version.
+func (p *Protocol) APIForVersion(version iotago.Version) (api iotago.API, err error) {
+	if mainEngineInstance := p.Engines.Main.Get(); mainEngineInstance != nil {
+		return mainEngineInstance.APIForVersion(version)
+	}
 
-	return p.mainEngine
+	return nil, ierrors.New("no engine instance available")
 }
 
-func (p *Protocol) CandidateEngineInstance() *engine.Engine {
-	p.activeEngineMutex.RLock()
-	defer p.activeEngineMutex.RUnlock()
-
-	if p.candidateEngine == nil {
-		return nil
-	}
+// APIForSlot returns the API for the given slot.
+func (p *Protocol) APIForSlot(slot iotago.SlotIndex) iotago.API {
+	return p.Engines.Main.Get().APIForSlot(slot)
+}
 
-	return p.candidateEngine.engine
+// APIForEpoch returns the API for the given epoch.
+func (p *Protocol) APIForEpoch(epoch iotago.EpochIndex) iotago.API {
+	return p.Engines.Main.Get().APIForEpoch(epoch)
 }
 
-func (p *Protocol) Network() *core.Protocol {
-	return p.networkProtocol
+// APIForTime returns the API for the given time.
+func (p *Protocol) APIForTime(t time.Time) iotago.API {
+	return p.Engines.Main.Get().APIForTime(t)
 }
 
+// CommittedAPI returns the API for the committed state.
 func (p *Protocol) CommittedAPI() iotago.API {
-	return p.MainEngineInstance().CommittedAPI()
+	return p.Engines.Main.Get().CommittedAPI()
 }
 
+// LatestAPI returns the latest API.
 func (p *Protocol) LatestAPI() iotago.API {
-	return p.MainEngineInstance().LatestAPI()
+	return p.Engines.Main.Get().LatestAPI()
+}
+
+// initSubcomponents initializes the subcomponents of the protocol and returns a function that shuts them down.
+func (p *Protocol) initSubcomponents(networkEndpoint network.Endpoint) (shutdown func()) {
+	p.Network = core.NewProtocol(networkEndpoint, p.Workers.CreatePool("NetworkProtocol"), p)
+	p.BlocksProtocol = newBlocksProtocol(p)
+	p.CommitmentsProtocol = newCommitmentsProtocol(p)
+	p.AttestationsProtocol = newAttestationsProtocol(p)
+	p.WarpSyncProtocol = newWarpSyncProtocol(p)
+	p.Commitments = newCommitments(p)
+	p.Chains = newChains(p)
+	p.Engines = newEngines(p)
+
+	return func() {
+		p.BlocksProtocol.Shutdown()
+		p.CommitmentsProtocol.Shutdown()
+		p.AttestationsProtocol.Shutdown()
+		p.WarpSyncProtocol.Shutdown()
+		p.Network.Shutdown()
+		p.Workers.WaitChildren()
+		p.Engines.Shutdown.Trigger()
+		p.Workers.Shutdown()
+	}
 }
 
-func (p *Protocol) APIForVersion(version iotago.Version) (iotago.API, error) {
-	return p.MainEngineInstance().APIForVersion(version)
+// initEviction initializes the eviction of old data when the engine advances and returns a function that shuts it down.
+func (p *Protocol) initEviction() (shutdown func()) {
+	return p.Commitments.Root.OnUpdate(func(_ *Commitment, rootCommitment *Commitment) {
+		// TODO: DECIDE ON DATA AVAILABILITY TIMESPAN / EVICTION STRATEGY
```

Comment:
This eviction is only for the protocol and its components like commitments and chains etc?

Reply:
yes